mixer/pkg/cache/cache.go

```diff
@@ -24,32 +24,49 @@ import (
 // Stats returns usage statistics about an individual cache, useful to assess the
 // efficiency of a cache.
 //
-// The values returned in this struct are eventually consistent approximations of
-// the current state of the cache.
+// The values returned in this struct are approximations of the current state of the cache.
+// For the sake of efficiency, certain edge cases in the implementation can lead to
+// inaccuracies.
 type Stats struct {
-	// Writes captures the number of times state in the cache was mutated. This
-	// includes setting and removing entries.
+	// Writes captures the number of times state in the cache was added or updated.
 	Writes uint64
 
 	// Hits captures the number of times a Get operation succeeded to find an entry in the cache.
 	Hits uint64
 
 	// Misses captures the number of times a Get operation failed to find an entry in the cache.
 	Misses uint64
+
+	// Evictions captures the number of entries that have been evicted from the cache
+	Evictions uint64
+
+	// Removals captures the number of entries that have been explicitly removed from the
+	// cache
+	Removals uint64
 }
 
-// Cache defines the standard behavior of in-memory caches.
+// Cache defines the standard behavior of in-memory thread-safe caches.
 //
 // Different caches can have different eviction policies which determine
 // when and how entries are automatically removed from the cache.
 //
-// Ideas for the future:
-//   - Return the number of entries in the cache in stats.
-//   - Provide an eviction callback to know when entries are evicted.
-//   - Have Set and Remove return the previous value for the key, if any.
-//   - Have Get return the expiration time for entries.
-//   - Add a Clear method to empty a cache.
+// Using a cache is very simple:
+//
+//   c := NewLRU(5*time.Second, 5*time.Second, 500)
+//   c.Set("foo", "bar")			// add an entry
+//   value, ok := c.Get("foo")		// try to retrieve the entry
+//   if ok {
+//		fmt.Printf("Got value %v\n", value)
+//   } else {
+//      fmt.Printf("Value was not found, must have been evicted")
+//   }
 type Cache interface {
+	// Ideas for the future:
+	//   - Return the number of entries in the cache in stats.
+	//   - Provide an eviction callback to know when entries are evicted.
+	//   - Have Set and Remove return the previous value for the key, if any.
+	//   - Have Get return the expiration time for entries.
```

Comment:
non-blocking: did we ever want to track promotion in LRU on Gets handled in background to Ideas?

Reply:
Added as a comment.